pkg/operation/botanist/controlplane/kubecontrollermanager/waiter_test.go

```diff
@@ -0,0 +1,244 @@
+)
+
+var _ = Describe("WaiterTest", func() {
+	var (
+		ctx                   = context.TODO()
+		testLogger            = logrus.NewEntry(logger.NewNopLogger())
+		errorMsg              = "fake error"
+		fakeErr               = fmt.Errorf(errorMsg)
+		kubeControllerManager KubeControllerManager
+		namespace             = "shoot--foo--bar"
+
+		// mock
+		ctrl        *gomock.Controller
+		seedClient  *mockclient.MockClient
+		shootClient *mockclient.MockClient
+		waiter      *retryfake.Ops
+		cleanupFunc func()
+
+		listOptions = []client.ListOption{
+			client.InNamespace(namespace),
+			client.MatchingLabels(map[string]string{
+				"app":  "kubernetes",
+				"role": "controller-manager",
+			}),
+		}
+	)
+
+	BeforeEach(func() {
+		ctrl = gomock.NewController(GinkgoT())
+		seedClient = mockclient.NewMockClient(ctrl)
+		shootClient = mockclient.NewMockClient(ctrl)
+	})
+
+	AfterEach(func() {
+		ctrl.Finish()
+		cleanupFunc()
+	})
+
+	Describe("#WaitForControllerToBeActive", func() {
+		BeforeEach(func() {
+			kubeControllerManager = New(
+				testLogger,
+				seedClient,
+				namespace,
+				nil,
+				"",
+				nil,
+				nil,
+				nil,
+			)
+
+			kubeControllerManager.SetShootClient(shootClient)
+
+			waiter = &retryfake.Ops{MaxAttempts: 1}
+			cleanupFunc = test.WithVars(
+				&retry.Until, waiter.Until,
+				&retry.UntilTimeout, waiter.UntilTimeout,
+			)
+		})
+
+		It("should fail if the seed client cannot talk to the Seed API Server", func() {
+			gomock.InOrder(
+				seedClient.EXPECT().Get(ctx, kutil.Key(namespace, "kube-controller-manager"), gomock.AssignableToTypeOf(&appsv1.Deployment{})).Return(fakeErr),
+			)
+
+			Expect(kubeControllerManager.WaitForControllerToBeActive(ctx)).To(MatchError(fakeErr))
+		})
+
+		It("should fail if the kube controller manager deployment does not exist", func() {
+			notFoundError := apierrors.NewNotFound(schema.GroupResource{}, "kube-controller-manager")
+			gomock.InOrder(
+				seedClient.EXPECT().Get(ctx, kutil.Key(namespace, "kube-controller-manager"), gomock.AssignableToTypeOf(&appsv1.Deployment{})).Return(notFoundError),
+			)
+
+			err := kubeControllerManager.WaitForControllerToBeActive(ctx)
+			Expect(err).To(HaveOccurred())
+			Expect(err).To(MatchError("kube controller manager deployment not found:  \"kube-controller-manager\" not found"))
+		})
+
+		It("should fail if it fails to list pods in the shoot namespace in the Seed", func() {
+			gomock.InOrder(
+				seedClient.EXPECT().Get(ctx, kutil.Key(namespace, "kube-controller-manager"), gomock.AssignableToTypeOf(&appsv1.Deployment{})),
+				seedClient.EXPECT().List(gomock.Any(), gomock.AssignableToTypeOf(&corev1.PodList{}), listOptions).Return(fakeErr),
+			)
+
+			err := kubeControllerManager.WaitForControllerToBeActive(ctx)
+			Expect(err).To(MatchError(fmt.Sprintf("could not check whether controller kube-controller-manager is active: %s", errorMsg)))
+		})
+
+		It("should fail if there is more than one kube controller manager pod", func() {
```

Comment:
Maybe add a test if there are no pods returned (but deployment exists)?

Reply:
That is a very good idea. This will prevent successful reconciliations if the watchdog downscaled the KCM pod.

Reply:
Added a test that should fail if no kube controller manager pod can be found